Client/Frontend/Browser/Helpers/LaunchHelper.swift

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveCore
+
+/// This class helps to prepare the browser during launch by ensuring the state of managers, resources and downloaders before performing additional tasks.
+public actor LaunchHelper {
+  public static let shared = LaunchHelper()
+  private var loadTask: Task<(), Never>?
+  private var areAdBlockServicesReady = false
+  
+  /// This method prepares the ad-block services one time so that multiple scenes can benefit from its results
+  /// This is particularly important since we use a shared instance for most of our ad-block services.
+  public func prepareAdBlockServices(adBlockService: AdblockService) async {
+    // Check if ad-block services are already ready.
+    // If so, we don't have to do anything
+    guard !areAdBlockServicesReady else { return }
+    
+    // Check if we're still preparing the ad-block services
+    // If so we await that task
+    if let task = loadTask {
+      return await task.value
+    }
+    
+    // Otherwise prepare the services and await the task
+    let task = Task { @MainActor in
+      // Load cached data
+      // This is done first because compileResources and loadCachedRuleLists need their results
+      async let loadBundledResources: Void = await ContentBlockerManager.shared.loadBundledResources()
+      async let filterListCache: Void = await FilterListResourceDownloader.shared.loadCachedData()
+      async let adblockResourceCache: Void = await AdblockResourceDownloader.shared.loadCachedData()
+      _ = await (loadBundledResources, filterListCache, adblockResourceCache)
+
+      // Compile some engines and load cached rule lists
+      async let compiledResourcesCompile: Void = await AdBlockEngineManager.shared.compileResources()
+      async let cachedRuleListLoad: Void = await ContentBlockerManager.shared.loadCachedRuleLists()
+      _ = await (compiledResourcesCompile, cachedRuleListLoad)
+      
+      // This one is non-blocking
+      await performPostLoadTasks(adBlockService: adBlockService)
+      await markAdBlockReady()
+    }
+    
+    // Await the task and wait for the results
+    self.loadTask = task
+    await task.value
+    self.loadTask = nil
+  }
+  
+  private func markAdBlockReady() {
+    areAdBlockServicesReady = true
+  }
+  
+  private func performPostLoadTasks(adBlockService: AdblockService) {
+    Task { @MainActor in
+      await ContentBlockerManager.shared.cleanupDeadRuleLists()
+      await ContentBlockerManager.shared.compilePendingResources()
```

Comment:
Can these be ran in parallel, or is the ordering here strict?

Reply:
I think so, the one removes any rules that shouldn't be there the other one compiles pending one. Since pending ones are a subset of enabled resources the two shouldn't interfere

Reply:
But it shouldn't affect load time because this is not blocking. It may affect however how quickly we get the latest rules in.

Reply:
We should run them async then, for better performance and to free up the main-thread :)

Reply:
Let's handle separately https://github.com/brave/brave-ios/issues/6423

Reply:
Oops, I accedentally resolved this one. @Brandon-T ill let you resolve this one if you agree